File: platform/polyglot/rag_embedding_config/py/rag_embedding_config/get_proxy.py
"""Async proxy URL resolution for OpenAI-compatible embedding providers.

Centralizes proxy lookup so it can be awaited from any transport builder.
Resolution order:

1. YAML ``providers.openai_embeddings.proxy_url``
   - ``false`` → disabled (return None)
   - string   → use it
   - ``null``  → inherit from global
2. YAML ``global.network.proxy_urls.{APP_ENV}``
3. ``OPENAI_EMBEDDINGS_PROXY_URL`` env var
4. None
"""

from __future__ import annotations

import os

from env_resolver import resolve_openai_env

_openai_env = resolve_openai_env()

_TAG = "[embedding-kwargs-openai]"


async def get_proxy(yaml_cfg=None) -> str | None:
    """Return the resolved proxy URL for OpenAI embeddings.

    Parameters
    ----------
    yaml_cfg:
        Optional YAML config object (AppYamlConfig instance or resolved dict).
        When ``None`` the YAML layer is skipped and only env vars are checked.
    """
    from .kwargs_openai import _yaml_get

    return _resolve(yaml_cfg, _yaml_get)


def get_proxy_sync(yaml_cfg=None) -> str | None:
    """Synchronous variant of :func:`get_proxy`."""
    from .kwargs_openai import _yaml_get

    return _resolve(yaml_cfg, _yaml_get)


def _resolve(yaml_cfg, yaml_get) -> str | None:
    """Shared resolution logic."""
    provider_proxy = yaml_get(yaml_cfg, "providers", "openai_embeddings", "proxy_url")

    if provider_proxy is not None:
        # Explicit false means "no proxy"
        if provider_proxy is False:
            print(f"{_TAG}   proxy: disabled (provider=false)")
            return None
        # Explicit string
        if isinstance(provider_proxy, str) and provider_proxy:
            print(f"{_TAG}   proxy: {provider_proxy} (provider)")
            return provider_proxy

    # Inherit from global network config
    app_env = os.environ.get("APP_ENV", "dev").lower()
    global_proxy = yaml_get(yaml_cfg, "global", "network", "proxy_urls", app_env)
    if global_proxy:
        print(f"{_TAG}   proxy: {global_proxy} (global.network.proxy_urls.{app_env})")
        return global_proxy

    # Env var fallback
    env_proxy = _openai_env.embeddings_proxy_url
    if env_proxy:
        print(f"{_TAG}   proxy: {env_proxy} (env)")
        return env_proxy

    print(f"{_TAG}   proxy: (none)")
    return None

File: platform/polyglot/rag_embedding_config/mjs/src/get_proxy.mjs
/**
 * @fileoverview Async proxy URL resolution for OpenAI-compatible embedding providers.
 *
 * Centralizes proxy lookup so it can be awaited from any transport builder.
 * Resolution order:
 *
 *   1. YAML `providers.openai_embeddings.proxy_url`
 *      - `false` → disabled (return undefined)
 *      - string  → use it
 *      - `null`  → inherit from global
 *   2. YAML `global.network.proxy_urls.{APP_ENV}`
 *   3. `OPENAI_EMBEDDINGS_PROXY_URL` env var
 *   4. undefined
 */

import { resolveOpenaiEnv } from '@internal/env-resolver';

const _openaiEnv = resolveOpenaiEnv();

const _TAG = '[embedding-kwargs-openai]';

/**
 * Return the resolved proxy URL for OpenAI embeddings.
 *
 * @param {object|null} [yamlCfg] - Optional YAML config object.
 * @param {(cfg: object|null, keys: string[], defaultValue?: *) => *} [yamlGet] - YAML getter function.
 * @returns {Promise<string|undefined>}
 */
export async function getProxy(yamlCfg = null, yamlGet = _defaultYamlGet) {
  return _resolve(yamlCfg, yamlGet);
}

/**
 * Synchronous variant of {@link getProxy}.
 *
 * @param {object|null} [yamlCfg]
 * @param {(cfg: object|null, keys: string[], defaultValue?: *) => *} [yamlGet]
 * @returns {string|undefined}
 */
export function getProxySync(yamlCfg = null, yamlGet = _defaultYamlGet) {
  return _resolve(yamlCfg, yamlGet);
}

/**
 * Shared resolution logic.
 * @param {object|null} yamlCfg
 * @param {(cfg: object|null, keys: string[], defaultValue?: *) => *} yamlGet
 * @returns {string|undefined}
 */
function _resolve(yamlCfg, yamlGet) {
  const providerProxy = yamlGet(yamlCfg, ['providers', 'openai_embeddings', 'proxy_url']);

  if (providerProxy !== undefined && providerProxy !== null) {
    if (providerProxy === false) {
      console.log(`${_TAG}   proxy: disabled (provider=false)`);
      return undefined;
    }
    if (typeof providerProxy === 'string' && providerProxy) {
      console.log(`${_TAG}   proxy: ${providerProxy} (provider)`);
      return providerProxy;
    }
  }

  // Inherit from global network config
  const appEnv = (process.env.APP_ENV || 'dev').toLowerCase();
  const globalProxy = yamlGet(yamlCfg, ['global', 'network', 'proxy_urls', appEnv]);
  if (globalProxy) {
    console.log(`${_TAG}   proxy: ${globalProxy} (global.network.proxy_urls.${appEnv})`);
    return globalProxy;
  }

  // Env var fallback
  const envProxy = _openaiEnv.embeddingsProxyUrl;
  if (envProxy) {
    console.log(`${_TAG}   proxy: ${envProxy} (env)`);
    return envProxy;
  }

  console.log(`${_TAG}   proxy: (none)`);
  return undefined;
}

/**
 * Minimal fallback yaml getter when no custom one is provided.
 * @param {object|null} cfg
 * @param {string[]} keys
 * @param {*} [defaultValue]
 * @returns {*}
 */
function _defaultYamlGet(cfg, keys, defaultValue = undefined) {
  if (!cfg) return defaultValue;
  if (typeof cfg.getNested === 'function') {
    return cfg.getNested(keys, defaultValue) ?? defaultValue;
  }
  let current = cfg;
  for (const key of keys) {
    if (current && typeof current === 'object' && key in current) {
      current = current[key];
    } else {
      return defaultValue;
    }
  }
  return current;
}
